aws/resource_aws_iot_thing_type.go

```diff
@@ -0,0 +1,186 @@
+		Create: resourceAwsIotThingTypeCreate,
+		Read:   resourceAwsIotThingTypeRead,
+		Update: resourceAwsIotThingTypeUpdate,
+		Delete: resourceAwsIotThingTypeDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				d.Set("name", d.Id())
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateIotThingTypeName,
+			},
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateIotThingTypeDescription,
+			},
+			"searchable_attributes": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				MaxItems: 3,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validateIotThingTypeSearchableAttribute,
+				},
+			},
+			"deprecated": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsIotThingTypeCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	params := &iot.CreateThingTypeInput{
+		ThingTypeName: aws.String(d.Get("name").(string)),
+		ThingTypeProperties: &iot.ThingTypeProperties{
+			SearchableAttributes: expandStringList(d.Get("searchable_attributes").(*schema.Set).List()),
+			ThingTypeDescription: aws.String(d.Get("description").(string)),
+		},
+	}
+
+	log.Printf("[DEBUG] Creating IoT Thing Type: %s", params)
+	out, err := conn.CreateThingType(params)
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.ThingTypeName)
+
+	return resourceAwsIotThingTypeUpdate(d, meta)
+}
+
+func resourceAwsIotThingTypeRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	params := &iot.DescribeThingTypeInput{
+		ThingTypeName: aws.String(d.Id()),
+	}
+	log.Printf("[DEBUG] Reading IoT Thing Type: %s", params)
+	out, err := conn.DescribeThingType(params)
+
+	if err != nil {
+		if isAWSErr(err, iot.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] IoT Thing Type %q not found, removing from state", d.Id())
+			d.SetId("")
+		}
+		return err
+	}
+
+	if out.ThingTypeMetadata != nil {
+		d.Set("deprecated", out.ThingTypeMetadata.Deprecated)
+	}
+
+	d.Set("arn", *out.ThingTypeArn)
+	properties := out.ThingTypeProperties
+	if properties != nil {
+		d.Set("searchable_attributes", flattenStringList(properties.SearchableAttributes))
+
+		if properties.ThingTypeDescription != nil {
+			d.Set("description", *properties.ThingTypeDescription)
+		}
+	}
+
+	return nil
+}
+
+func resourceAwsIotThingTypeUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	if d.HasChange("deprecated") {
```

Comment:
Unless I'm mistaken this will always be triggered for new resources during creation, even though it's IMO not necessary as we can assume new resource doesn't need "undeprecation" (no-op API call)?

Reply:
Theoretically we could just modify the condition here to also check for `d.IsNewResource()` and the real value of `deprecated`.